com.unity.render-pipelines.core/Runtime/Volume/KeyframeUtility.cs

```diff
@@ -0,0 +1,294 @@
+    /// InterpAnimationCurve calculates a synthetic keypoint at that time based on value and derivative, and interpolates
+    /// the resulting keys.
+    ///
+    /// Note that this function should only be called by internal rendering code. It creates a small pool of animation
+    /// curves and reuses them to avoid creating garbage. The number of curves needed is quite small, since curves only need
+    /// to be used when interpolating multiple volumes together with different curve parameters. The underlying interp
+    /// function isn't allowed to fail, so in the case where we run out of memory we fall back to returning a single keyframe.
+    /// </summary>
+    public class KeyframeUtility
+    {
+        /// <summary>
+        /// Helper function to remove all control points for an animation curve. Since animation curves are reused in a pool,
+        /// this function clears existing keys so the curve is ready for reuse.
+        /// </summary>
+        /// <param name="curve">The curve to reset.</param>
+        static public void ResetAnimationCurve(ref AnimationCurve curve)
+        {
+            int numPoints = curve.length;
+            for (int i = numPoints - 1; i >= 0; i--)
+            {
+                curve.RemoveKey(i);
+            }
+        }
+
+        static private Keyframe LerpSingleKeyframe(Keyframe lhs, Keyframe rhs, float t)
+        {
+            var ret = new Keyframe();
+
+            ret.time = Mathf.Lerp(lhs.time, rhs.time, t);
+            ret.value = Mathf.Lerp(lhs.value, rhs.value, t);
+            ret.inTangent = Mathf.Lerp(lhs.inTangent, rhs.inTangent, t);
+            ret.outTangent = Mathf.Lerp(lhs.outTangent, rhs.outTangent, t);
+            ret.inWeight = Mathf.Lerp(lhs.inWeight, rhs.inWeight, t);
+            ret.outWeight = Mathf.Lerp(lhs.outWeight, rhs.outWeight, t);
+
+            // it's not possible to lerp the weightedMode, so use the lhs mode.
+            ret.weightedMode = lhs.weightedMode;
+
+            // Note: ret.tangentMode is deprecated, so we will use  the value from the constructor
+            return ret;
+        }
+
+        /// In an animation curve, the inTangent and outTangent don't match the edge of the curve. For example,
+        /// the first key might have inTangent=3.0f but the actual incoming tangent is 0.0 because the curve is
+        /// clamped outside the time domain. So this helper fetches a key, but zeroes out the inTangent of the first
+        /// key and the outTangent of the last key.
+        static private Keyframe GetKeyframeAndClampEdge([DisallowNull] Keyframe[] keys, int index)
+        {
+            var currKey = keys[index];
+            if (index == 0)
+            {
+                currKey.inTangent = 0.0f;
+            }
+            if (index == keys.Length - 1)
+            {
+                currKey.outTangent = 0.0f;
+            }
+            return currKey;
```

Comment:
```suggestion
            var keysSize = keys.Length -1:
            if (index < 0 || index > keysSize)
               return new KeyFrame();
               
            var currKey = keys[index];
            if (index == 0)
            {
                currKey.inTangent = 0.0f;
            }
            else if (index == keysSize)
            {
                currKey.outTangent = 0.0f;
            }
            return currKey;
```

I think we should think about this, or either return null but this would mean that code handling this should handle null.

Reply:
The concern is that returning invalid data (due to a programmer bug) is often worse than stopping right away (either through an assert() or an OOB exception). Suppose a programmer introduces a bug, and we silently return an empty keyframe. The user likely wouldn't know, and we would end up with a very difficult bug to track down. I.e. at some point a QA person realizes that there is a small stutter when transitioning between volumes.

Personally, I don't have any strong feelings on the actual code structure. But it is important that if we have a bug, that we can detect it right away instead of silently hiding it. So I'm including your change, but also adding a debug warning to the log before the return statement.